proof/clear_cofactor_p2.saw

```diff
@@ -0,0 +1,210 @@
+  do {
+    unfolding ["POINTonE2_times_minus_z"];
+    simplify (addsimps (foldr concat [POINTonE2_thms, e2_curve_op_thms, e2_scalar_mult_alg_thms] []) empty_ss);
+    w4_unint_z3 ["affinify_E'", "E'", "POINTonE2_abs", "e2_scalar_mult"];
+  };
+
+// Prove times_minus_z is equivalent to e2_scalar_mult
+times_minus_z_thm <- custom_prove_cryptol
+  {{ \P -> affinify E' (times_minus_z P) == e2_scalar_mult pos_c1 (affinify E' P) }}
+  do {
+    simplify (addsimps [times_minus_z_rev_thm, affinify_POINTonE2_times_minus_z_thm, affinify_E'_thm] empty_ss);
+    simplify (addsimps POINTonE2_thms empty_ss);
+    w4_unint_z3 ["e2_scalar_mult", "affinify_E'"];
+  };
+
+// Algebraic theorems
+sub_to_add_neg_p2_thm <- prove_cryptol
+  {{ \p1 p2 -> sub E' p1 p2 == add E' p1 (neg E' p2) }} ["E'"];
+
+psi_neg_thm <- prove_cryptol {{ \P -> psi (neg E' P) == neg E' (psi P) }} [];
+
+neg_neg_p2_thm <- prove_cryptol
+  {{ \P -> neg E' (neg E' P) == (P.0, Fp_2.normalize P.1)}} [];
+
+mult_commutes_p2_thm <- admit_cryptol
+  {{ \x y z -> e2_scalar_mult x (add E' y z) == add E' (e2_scalar_mult x y) (e2_scalar_mult x z) }};
+
+pull_mult_neg_p2_thm <- admit_cryptol
+  {{ \x y -> e2_scalar_mult x (neg E' y) == neg E' (e2_scalar_mult x y) }};
+
+push_neg_add_p2_thm <- admit_cryptol
+  {{ \x y -> neg E' (add E' x y) == add E' (neg E' x) (neg E' y) }};
+
+let add_sort_p2_laws = [
+    {{ \ x y   -> add_E' x y == add_E' y x }} // add commutes
+  , {{ \ x y z -> add_E' (add_E' x y) z == add_E' x (add_E' y z) }} // add associates
+  , {{ \ x y z -> add_E' x (add_E' y z) == add_E' y (add_E' x z) }} // add commutes, again
+  , {{ \ b x1 x2 y -> add_E' (if b then x1 else x2) y ==
+                      if b then add_E' x1 y else add_E' x2 y }}
+];
+
+add_sort_p2_thms <- for add_sort_p2_laws (\t -> admit_cryptol t);
+
+e2_scalar_mult_thm <- admit_cryptol {{  \x y -> scalar_mult E' x y == e2_scalar_mult x y }};
+
+let {{
+  AffinePoint_Fp_2_invariant: AffinePoint t_Fp_2 -> Bool
+  AffinePoint_Fp_2_invariant (x, y) = Fp_2.is_val x /\ Fp_2.is_val y
+  }};
+
+Fp_2_normalize_second_thm <- prove_cryptol
+  {{ \P -> (P.0, Fp_2.normalize P.1) ==
+           if AffinePoint_Fp_2_invariant P
+           then P
+           else (P.0, apply Fp_2.normalize P.1) }} [];
+
+e2_scalar_mult_inv_thm <- admit_cryptol
+  {{ \P -> AffinePoint_Fp_2_invariant (e2_scalar_mult pos_c1 P) ==
+           if AffinePoint_Fp_2_invariant P
+           then True
+           else apply AffinePoint_Fp_2_invariant (e2_scalar_mult pos_c1 P) }};
+
+// Prove clear_cofactor_bls12381_g2 (RFC spec) is equivalent to
+// clear_cofactor_E2_impl (spec based on C clear_cofactor_E2 function).
+clear_cofactor_E2_equiv_thm <- custom_prove_cryptol
+  {{ \P -> if AffinePoint_Fp_2_invariant (affinify E' P)
+           then clear_cofactor_bls12381_g2 (affinify E' P) ==
+                affinify E' (clear_cofactor_E2_impl P)
+           else True }}
```

Comment:
To use it as a rewrite rule, it would be easier to have
```
  {{ \P -> clear_cofactor_bls12381_g2 (affinify E' P) ==
      if AffinePoint_Fp_2_invariant (affinify E' P)
      then affinify E' (clear_cofactor_E2_impl P)
      else apply clear_cofactor_bls12381_g2 (affinify E' P) }}
```

Reply:
The commit I just pushed up changes the theorem to this form.